test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua

```diff
@@ -0,0 +1,69 @@
+local checktrace = require('jit.util').traceinfo
+local t = require('luatest')
+local g = t.group()
+
+-- Test payload to be recorded. It calculates the arithmetic sum
+-- of elements from 1 to the given n.
+local function sumN(n)
+    local result = 0
+    for i = 1, n do
+        result = result + i
+    end
+    return result
+end
+
+g.test_luajit_enable_checkhook = function()
+    -- Setup JIT engine to record a trace as soon as possible.
+    jit.opt.start('hotloop=1')
+    -- Purge all existing traces.
+    jit.flush()
+
+    -- Check there is no traces at the beginning of the test.
+    assert(not checktrace(1), "JIT flush doesn't work")
+    -- Run payload with the minimal possible number to force
+    -- the inner loop to be compiled.
+    sumN(3)
+    -- Check the new trace is compiled as a result of the payload
+    -- call above.
+    assert(checktrace(1), 'The loop in <sumN> is compiled')
+
+    -- Return JIT engine setup to its defaults.
+    jit.opt.start('hotloop=56')
+
+    -- Upvalues to fetch key and value from lines to be used in
+    -- the scope of the debug hook below.
+    local lidx, lval
+    -- Table with lines to be hit by debug hook below.
+    local lines = {
```

Comment:
Side note: I really hope that we will have no need to patch this tests (like we do for memprof).

Reply:
@fckxorg, nice, I'll rewrite the chunk this way.

Reply:
I have come up with the following change. If you have any better ideas, feel free to share them.

```diff
diff --git a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
index af48479df..242e3f8a1 100644
--- a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
+++ b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
@@ -4,13 +4,31 @@ local g = t.group()
 
 -- Test payload to be recorded. It calculates the arithmetic sum
 -- of elements from 1 to the given n.
-local function sumN(n)
-    local result = 0
-    for i = 1, n do
-        result = result + i
+-- XXX: To avoid another fragile test, use loadstring to generate
+-- the function with code motion agnostic line numbers to be used
+-- in hook validation below.
+local sumN = assert(load([[
+    return function(n)
+        local result = 0
+        for i = 1, n do
+            result = result + i
+        end
+        return result
     end
-    return result
-end
+]]))()
+
+-- Table with lines to be hit by debug hook below.
+local lines = {
+    2,  -- local result = 0
+    3,  -- for i = 1, n do
+    4,  --   result = result + i -- (i is 1)
+    3,  -- for i = 1, n do
+    4,  --   result = result + i -- (i is 2)
+    3,  -- for i = 1, n do
+    4,  --   result = result + i -- (i is 3)
+    3,  -- for i = 1, n do
+    6,  -- return result
+}
 
 g.test_luajit_enable_checkhook = function()
     -- Setup JIT engine to record a trace as soon as possible.
@@ -30,30 +48,34 @@ g.test_luajit_enable_checkhook = function()
     -- Return JIT engine setup to its defaults.
     jit.opt.start('hotloop=56')
 
-    -- Upvalues to fetch key and value from lines to be used in
-    -- the scope of the debug hook below.
-    local lidx, lval
-    -- Table with lines to be hit by debug hook below.
-    local lines = {
-        60, -- local sum = sumN(3)
-        8,  -- local result = 0
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 1)
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 2)
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 3)
-        9,  -- for i = 1, n do
-        12, -- return result
-        63, -- debug.sethook()
-    }
+    -- Upvalue for a key from lines to be used in the scope of the
+    -- the debug hook below.
+    local lidx = 1
+
-    -- Upvalues to fetch key and value from lines to be used in
-    -- the scope of the debug hook below.
-    local lidx, lval
-    -- Table with lines to be hit by debug hook below.
-    local lines = {
-        60, -- local sum = sumN(3)
-        8,  -- local result = 0
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 1)
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 2)
-        9,  -- for i = 1, n do
-        10, --   result = result + i -- (i is 3)
-        9,  -- for i = 1, n do
-        12, -- return result
-        63, -- debug.sethook()
-    }
+    -- Upvalue for a key from lines to be used in the scope of the
+    -- the debug hook below.
+    local lidx = 1
+
+    -- XXX: The only fragile part of this test is here. After the
+    -- hook is set few lines below, it is fired two more times in
+    -- addition to all entries in <lines> table above. Hence we
+    -- have to extend <lines> table with the following entries.
+    -- 1. The call of the compiled test payload below.
+    -- local sum = sumN(3)
+    -- In this case the hook is fired before any of the lines
+    -- within <sumN> function, so the entry is added at the least
+    -- possible index in the <lines> (i.e. <lidx> starting value).
+    table.insert(lines, lidx, 82)
+    -- 2. Turning of the hook after the test payload returns.
+    -- debug.sethook()
+    -- This is the last time the hook is fired, so the entry is
+    -- added to the end of the <lines> table.
+    table.insert(lines, 85)
 
     -- Set debug hook.
     debug.sethook(function(hookname, line)
```

Reply:
Finally I came up with the following: there is no need to check the first and the last entry of `lines` table, hence, just ignore the assertion when the hook is fired for these cases. The diff is below:
```diff
diff --git a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
index e3dea7c60..ee76bf7ee 100644
--- a/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
+++ b/test/app-luatest/gh_7762_luajit_enable_checkhook_test.lua
@@ -50,30 +50,32 @@ g.test_luajit_enable_checkhook = function()
 
     -- Upvalue for a key from lines to be used in the scope of the
     -- the debug hook below.
-    local lidx = 1
+    local lidx = 0
 
-    -- XXX: The only fragile part of this test is here. After the
-    -- hook is set few lines below, it is fired two more times in
-    -- addition to all entries in <lines> table above. Hence we
-    -- have to extend <lines> table with the following entries.
-    -- 1. The call of the compiled test payload below.
-    -- local sum = sumN(3)
-    -- In this case the hook is fired before any of the lines
-    -- within <sumN> function, so the entry is added at the least
-    -- possible index in the <lines> (i.e. <lidx> starting value).
-    table.insert(lines, lidx, 82)
-    -- 2. Turning of the hook after the test payload returns.
-    -- debug.sethook()
-    -- This is the last time the hook is fired, so the entry is
-    -- added to the end of the <lines> table.
-    table.insert(lines, 85)
-
-    -- Set debug hook.
+    -- XXX: After the hook is set, it is fired two more times in
+    -- addition to all entries in <lines> table above:
+    -- 1. The call of the compiled test payload below:
+    --         local sum = sumN(3)
+    --   In this case the hook is fired before any of the lines
+    --   within <sumN> function, so we can ignore it by setting
+    --   <lidx> initial value to zero. As a result the lookup
+    --   in <lines> table yields nil for this case.
+    -- 2. Turning of the hook after the test payload returns:
+    --         debug.sethook()
+    --   This is the last time the hook is fired, and <lidx> value
+    --   equals to #lines + 1. As a result the lookup also yields
+    --   nil for this case.
+    -- Since the test is aimed to check whether the hook is fired
+    -- on trace compiled for the payload above, there is no need
+    -- to check the actual value of the line for the both
+    -- aforementioned cases and we can simply skip the assertion.
     debug.sethook(function(hookname, line)
         -- XXX: Only line hooks are fired.
         assert(hookname == 'line', 'Unexpected hook type is fired')
         local lval = lines[lidx]
-        t.assert_equals(line, lval)
+        if lval then
+            t.assert_equals(line, lval)
+        end
         lidx = lidx + 1
     end, 'l')
```